STM32/ACTenChannel/Core/Src/DS18B20.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * DS18B20.c
+ *
+ *  Created on: 7 Nov 2024
+ *      Author: matias
+ * 
+ * 	Interface based heavily on the implementation done by [vtx22](https://github.com/vtx22/STM32-DS18B20/blob/master/src/DS18B20.cpp)
+ *  Detailed explanation of DS18B20 communication interface: https://controllerstech.com/ds18b20-and-stm32/
+ *  Datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/DS18B20.pdf
+ */
+
+#include "DS18B20.h"
+
+TIM_HandleTypeDef* DS18B20_tim = NULL;
+
+typedef struct Comm {
+    StmGpio * dataGpio;	// GPIO for sending/receiving data to/from DS18B20
+    GPIO_TypeDef *blk; 	
+    uint16_t pin;
+} Comm; 
+
+Comm comm = {0};
+
+/***************************************************************************************************
+** PRIVATE FUNCTION PROTOTYPES
+***************************************************************************************************/
+
+static void delay_us(uint32_t us);
+static void writeData(uint8_t data);
+static uint8_t readData();
+static uint8_t startSensor();
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+void setPinOutput()
+{
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+    GPIO_InitStruct.Pin = comm.pin;
+    GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
+    GPIO_InitStruct.Speed = GPIO_SPEED_HIGH;
+    HAL_GPIO_Init(comm.blk, &GPIO_InitStruct);
+    stmGpioInit(comm.dataGpio, comm.blk, comm.pin, STM_GPIO_OUTPUT);
+}
+
+void setPinInput()
+{
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+    GPIO_InitStruct.Pin = comm.pin;
+    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
+    GPIO_InitStruct.Speed = GPIO_SPEED_HIGH;
+    GPIO_InitStruct.Pull = GPIO_PULLUP;
+    HAL_GPIO_Init(comm.blk, &GPIO_InitStruct);
+    stmGpioInit(comm.dataGpio, comm.blk, comm.pin, STM_GPIO_INPUT);
+}
+
+/*!
+** @brief Returns after us microseconds
+** @note The DS18B20_tim updates with 1MHz
+*/
+static void delay_us(uint32_t us)
+{
+    __HAL_TIM_SET_COUNTER(DS18B20_tim, 0);
+    while (DS18B20_tim->Instance->CNT < us);
+}
+
+/*!
+** @brief Sends a byte to the DS18B20
+*/
+static void writeData(uint8_t data)
+{
+    for (uint8_t i = 0; i < 8; i++)
+    {
+        if (data & (1 << i))
+        {
+            setPinOutput();
+            stmSetGpio(*comm.dataGpio, false);
+            delay_us(1);
+
+            setPinInput();
+            delay_us(50);
+            continue;
+        }
+
+        setPinOutput();
+        stmSetGpio(*comm.dataGpio, false);
+        delay_us(50);
+
+        setPinInput();
+    }
+}
+
+/*!
+** @brief Receives a byte from the DS18B20
+*/
+static uint8_t readData()
+{
+    uint8_t value = 0;
+    setPinInput();
```

Comment:
I don't think this line is needed as you make it an output on the next line

Reply:
All the other implementations I've seen has had this seemingly redundant call. I will try to remove it and see if it still works :=)